File: ios-bookmind/Sources/Retrieval/RetrievalEngine.swift
import Foundation
import SharedModels

public final class RetrievalEngine: ContextRetrieving {
    private let characters: CharacterLookupService
    private let facts: FactSearchService
    private let chunks: ChunkSearchService
    private let recaps: RecapLookupService
    private let events: EventLookupService

    public init(
        characters: CharacterLookupService,
        facts: FactSearchService,
        chunks: ChunkSearchService,
        recaps: RecapLookupService,
        events: EventLookupService
    ) {
        self.characters = characters
        self.facts = facts
        self.chunks = chunks
        self.recaps = recaps
        self.events = events
    }

    public func context(
        for question: String,
        bookID: BookID,
        currentChapterIndex: Int,
        spoilerBoundary: Int
    ) async throws -> RetrievalContext {
        let maxChapter = min(spoilerBoundary, currentChapterIndex)

        async let recapTask = recaps.latestRecap(bookID: bookID, maxChapterIndex: maxChapter)
        async let charactersTask = characters.findCharacters(
            matching: question,
            bookID: bookID,
            currentChapterIndex: maxChapter
        )
        async let factsTask = facts.searchFacts(
            query: question,
            bookID: bookID,
            maxChapterIndex: maxChapter
        )
        async let chunksTask = chunks.searchChunks(
            query: question,
            bookID: bookID,
            maxChapterIndex: maxChapter
        )
        async let eventsTask = events.recentEvents(
            bookID: bookID,
            maxChapterIndex: maxChapter,
            limit: 5
        )

        let (recap, chars, fts, chnks, evs) = try await (
            recapTask, charactersTask, factsTask, chunksTask, eventsTask
        )

        return rerank(
            RetrievalContext(
                safeRecap: recap,
                characterCards: chars,
                recentEvents: evs,
                quotes: chnks,
                facts: fts
            ),
            currentChapterIndex: currentChapterIndex,
            spoilerBoundary: maxChapter
        )
    }

    /// Drops anything beyond spoiler boundary (defense-in-depth) and
    /// sorts by proximity to current chapter / spoiler level.
    private func rerank(
        _ ctx: RetrievalContext,
        currentChapterIndex: Int,
        spoilerBoundary: Int
    ) -> RetrievalContext {
        let filteredCharacters = ctx.characterCards
            .filter { $0.firstChapterIndex <= spoilerBoundary }
            .sorted { lhs, rhs in
                let lDistance = abs(currentChapterIndex - lhs.firstChapterIndex)
                let rDistance = abs(currentChapterIndex - rhs.firstChapterIndex)
                return lDistance < rDistance
            }

        let filteredFacts = ctx.facts
            .filter { $0.chapterIndex <= spoilerBoundary && $0.spoilerLevel <= 1 }
            .sorted { lhs, rhs in
                if lhs.spoilerLevel != rhs.spoilerLevel { return lhs.spoilerLevel < rhs.spoilerLevel }
                return abs(currentChapterIndex - lhs.chapterIndex)
                    < abs(currentChapterIndex - rhs.chapterIndex)
            }

        let filteredChunks = ctx.quotes
            .filter { $0.spoilerLevel <= 1 }
            .sorted { abs(currentChapterIndex - $0.index) < abs(currentChapterIndex - $1.index) }

        let filteredEvents = ctx.recentEvents
            .filter { $0.chapterIndex <= spoilerBoundary && $0.spoilerLevel <= 1 }

        return RetrievalContext(
            safeRecap: ctx.safeRecap,
            characterCards: filteredCharacters,
            recentEvents: filteredEvents,
            quotes: filteredChunks,
            facts: filteredFacts
        )
    }
}

File: android-bookmind/app/src/main/java/com/bookmind/retrieval/RetrievalEngine.kt
package com.bookmind.retrieval

import com.bookmind.core.model.BookID
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import javax.inject.Inject
import kotlin.math.abs
import kotlin.math.min

/** = iOS `RetrievalEngine`. Gathers safe context concurrently, then reranks/filters. */
class RetrievalEngine @Inject constructor(
    private val characters: CharacterLookupService,
    private val facts: FactSearchService,
    private val chunks: ChunkSearchService,
    private val recaps: RecapLookupService,
    private val events: EventLookupService
) : ContextRetrieving {

    override suspend fun context(
        question: String,
        bookID: BookID,
        currentChapterIndex: Int,
        spoilerBoundary: Int
    ): RetrievalContext = coroutineScope {
        val maxChapter = min(spoilerBoundary, currentChapterIndex)

        val recapTask = async { recaps.latestRecap(bookID, maxChapter) }
        val charactersTask = async { characters.findCharacters(question, bookID, maxChapter) }
        val factsTask = async { facts.searchFacts(question, bookID, maxChapter) }
        val chunksTask = async { chunks.searchChunks(question, bookID, maxChapter) }
        val eventsTask = async { events.recentEvents(bookID, maxChapter, 5) }

        rerank(
            RetrievalContext(
                safeRecap = recapTask.await(),
                characterCards = charactersTask.await(),
                recentEvents = eventsTask.await(),
                quotes = chunksTask.await(),
                facts = factsTask.await()
            ),
            currentChapterIndex = currentChapterIndex,
            spoilerBoundary = maxChapter
        )
    }

    /** Drops anything past the spoiler boundary (defense-in-depth) and sorts by proximity. */
    private fun rerank(
        ctx: RetrievalContext,
        currentChapterIndex: Int,
        spoilerBoundary: Int
    ): RetrievalContext {
        val filteredCharacters = ctx.characterCards
            .filter { it.firstChapterIndex <= spoilerBoundary }
            .sortedBy { abs(currentChapterIndex - it.firstChapterIndex) }

        val filteredFacts = ctx.facts
            .filter { it.chapterIndex <= spoilerBoundary && it.spoilerLevel <= 1 }
            .sortedWith(
                compareBy({ it.spoilerLevel }, { abs(currentChapterIndex - it.chapterIndex) })
            )

        val filteredChunks = ctx.quotes
            .filter { it.spoilerLevel <= 1 }
            .sortedBy { abs(currentChapterIndex - it.index) }

        val filteredEvents = ctx.recentEvents
            .filter { it.chapterIndex <= spoilerBoundary && it.spoilerLevel <= 1 }

        return RetrievalContext(
            safeRecap = ctx.safeRecap,
            characterCards = filteredCharacters,
            recentEvents = filteredEvents,
            quotes = filteredChunks,
            facts = filteredFacts
        )
    }
}
